Corrections to TorqueReactionScalar in tensor mechanics

sapitts opened the pull request:
**Relevant Design Information**
Corrects an algorithm error in the scalar method to calculate the torque reaction in tensor mechanics, using [this method](http://www.maths.usyd.edu.au/u/MOW/vectors/vectors-10/v-10-1.html), and should fix the Intel compiler exodiff error.  To allow for 2D calculations, the aux variable names are treated as pointers and set to NULL if the corresponding reaction force is not specified in the input file.

**Test Cases**
The existing 3D torque reaction test case was re-golded, and an elastic strain 2D test case was added for quick verification with a statics problem.

Refs #6484

sapitts committed:
Corrections to TorqueReaction and DisplacementAboutAxis

migration with addition of 2D test for verification Refs #6484

sapitts commented:
@backmari and @bwspenc , would you please review this post processor as an eventual replacement for the solid mechanics TorqueReaction class?

bwspenc commented:
We just discussed this, and agreed upon the following changes:

- Rename the tensor_mechanics TorqueReactionScalar class to TorqueReaction
- Delete the solid_mechanics TorqueReaction class
- Make the x and y reactions required
- Unitize the axial direction vector in the solid_mechanics version and verify that the new implementation matches that, and pick one of the implementations once we verify that they match.

sapitts commented:
The updated changes include:
- A correction to the method used to compute the torque by projecting the cross of the moment arm and the force onto the user-specified axis direction vector
- Migration of the DisplacedAroundAxisBC and the addition of two more tests for the new BC
  - Set the abs_zero to a relatively high 5e-7 to avoid exodiff errors on small residual forces near the center of the cylinder in the torque_reaction_cylinder test case
- Deletion and not migration of the DisplaceAroundAxisAction

jwpeterson commented on modules/tensor_mechanics/src/bcs/DisplacementAboutAxis.C:
```diff
@@ -69,39 +70,46 @@ DisplacementAboutAxis::computeQpValue()
+  // only perform the operation if the provided direction vector is not already a unit vector
+  if (std::abs(1.0 - magnitude) > 1.0e-15)
+    for (unsigned int i = 0; i < 3; ++i)
+      _axis_direction(i) = _axis_direction(i) / magnitude;
```
Not sure if this could happen, but if `_axis_direction==(0,0,0)`, you would divide by zero here...

dschwen replied:
and why not just do 
```
_axis_direction /= magnitude;
```
instead of the manual looping over the components?

dschwen replied:
Is the abs test even necessary? Looks like premature optimization to me.

andrsd replied:
This seems possible, there is no check in c'tor that would prevent it. Let's add that check there...

sapitts replied:
I'll add in a test for the axis_direction (0,0,0); the direction vector needs to have some nonzero direction in order to calculate the torque.

Is c'tor constructor?

jwpeterson replied:
> Is c'tor constructor?

Yep.

sapitts commented:
Updated changes include:
- mooseError addition in the constructor to check for a zero direction vector
- removed tolerance in check for existing unit vector, but retained the check ```if (magnitude != 1)``` to maintain the Solid Mechanics capability
- Added two error checking tests for the displacement about axis BC

dschwen commented on modules/tensor_mechanics/src/bcs/DisplacementAboutAxis.C:
```diff
@@ -69,39 +73,45 @@ DisplacementAboutAxis::computeQpValue()
+  Real magnitude = _axis_direction.norm();
+
+  // only perform the normalization if the provided direction vector is not already a unit vector
+  if (magnitude != 1.0)
```
I still don't get how this check could ever be necessary. If magnitude == 1 then the next operation is a no-op.

sapitts replied:
Now that I've done some 'literature review' on no-op thanks to wikipedia, I suppose I can stop driving @dschwen crazy with this unnecessary check.
